File: sdk/typescript/test-websocket.ts
/**
 * YoorQuezt WebSocket Subscription Test
 *
 * Tests ws:subscribe scope — connects to gateway and subscribes to live streams.
 *
 * Usage:
 *   export YQ_API_KEY="yq_live_..."
 *   npx tsc && node dist/test-websocket.js
 */

import WebSocket from "ws";

const API_KEY = process.env.YQ_API_KEY;
if (!API_KEY) {
  console.error("Set YQ_API_KEY environment variable");
  process.exit(1);
}

const GATEWAY_WS = process.env.GATEWAY_WS || "wss://gateway-testnet.yoorquezt.io/ws";

const topics = ["auction", "mempool", "blocks", "protect", "intents"];

async function main() {
  console.log("==================================================");
  console.log("  YoorQuezt WebSocket Subscription Test");
  console.log(`  Gateway: ${GATEWAY_WS}`);
  console.log(`  Key:     ${API_KEY!.slice(0, 20)}...`);
  console.log("==================================================\n");

  const ws = new WebSocket(GATEWAY_WS);

  ws.on("open", () => {
    console.log("Connected to gateway\n");

    // Subscribe to all topics
    for (const topic of topics) {
      const msg = JSON.stringify({
        jsonrpc: "2.0",
        id: Date.now(),
        method: "mev_subscribe",
        params: { topic },
      });
      ws.send(msg);
      console.log(`  Subscribed to: ${topic}`);
    }

    console.log("\nListening for events (30s)...\n");
  });

  ws.on("message", (data) => {
    const msg = JSON.parse(data.toString());
    if (msg.method === "mev_subscription") {
      const topic = msg.params?.topic || "unknown";
      console.log(`  [${topic}] ${JSON.stringify(msg.params?.data || msg.params).slice(0, 200)}`);
    } else {
      console.log(`  [response] ${JSON.stringify(msg).slice(0, 200)}`);
    }
  });

  ws.on("error", (err) => {
    console.error(`  WebSocket error: ${err.message}`);
  });

  ws.on("close", (code, reason) => {
    console.log(`\nConnection closed (${code}): ${reason}`);
  });

  // Listen for 30 seconds
  await new Promise((resolve) => setTimeout(resolve, 30000));
  ws.close();
  console.log("\nDone.");
}

main().catch(console.error);

File: sdk/python/test_websocket.py
"""
YoorQuezt WebSocket Subscription Test (Python)

Tests ws:subscribe scope — connects to gateway and subscribes to live streams.

Usage:
    export YQ_API_KEY="yq_live_..."
    python test_websocket.py

Prerequisites:
    pip install websockets
"""

import os
import sys
import json
import asyncio
import ssl

try:
    import websockets
except ImportError:
    print("Install websockets: pip install websockets")
    sys.exit(1)

API_KEY = os.getenv("YQ_API_KEY")
if not API_KEY:
    print("Set YQ_API_KEY environment variable")
    sys.exit(1)

GATEWAY_WS = os.getenv("GATEWAY_WS", "wss://gateway-testnet.yoorquezt.io/ws")
TOPICS = ["auction", "mempool", "blocks", "protect", "intents"]


async def main():
    print("=" * 50)
    print("  YoorQuezt WebSocket Subscription Test")
    print(f"  Gateway: {GATEWAY_WS}")
    print(f"  Key:     {API_KEY[:20]}...")
    print("=" * 50)

    # Use default SSL context with certifi certs if available
    ssl_ctx = ssl.create_default_context()
    try:
        import certifi
        ssl_ctx.load_verify_locations(certifi.where())
    except ImportError:
        # Fall back to unverified if certifi not installed
        ssl_ctx.check_hostname = False
        ssl_ctx.verify_mode = ssl.CERT_NONE

    extra_headers = {"Authorization": f"Bearer {API_KEY}"}

    async with websockets.connect(
        GATEWAY_WS, ssl=ssl_ctx, additional_headers=extra_headers
    ) as ws:
        print("\nConnected to gateway\n")

        # Subscribe to all topics
        for i, topic in enumerate(TOPICS):
            msg = json.dumps({
                "jsonrpc": "2.0",
                "id": 100 + i,
                "method": "mev_subscribe",
                "params": {"topic": topic},
            })
            await ws.send(msg)
            print(f"  Subscribed to: {topic}")

        print("\nListening for events (30s)...\n")

        received = 0
        try:
            async with asyncio.timeout(30):
                async for message in ws:
                    data = json.loads(message)
                    if data.get("method") == "mev_subscription":
                        topic = data.get("params", {}).get("topic", "unknown")
                        payload = json.dumps(data.get("params", {}).get("data", data.get("params")))[:200]
                        print(f"  [{topic}] {payload}")
                        received += 1
                    else:
                        print(f"  [response] {json.dumps(data)[:200]}")
        except asyncio.TimeoutError:
            pass

    print(f"\nDone. Received {received} events.")


if __name__ == "__main__":
    asyncio.run(main())
